Source/Diagnostics/ReducedDiags/ChargeOnEB.cpp

```diff
@@ -0,0 +1,204 @@
+        std::string weighting_string = "";
+        utils::parser::Store_parserString(
+            pp_rd_name,"weighting_function(x,y,z)", weighting_string);
+        m_parser_weighting = std::make_unique<amrex::Parser>(
+            utils::parser::makeParser(weighting_string,{"x","y","z"}));
+    }
+
+    if (ParallelDescriptor::IOProcessor())
+    {
+        if ( m_IsNotRestart )
+        {
+            // open file
+            std::ofstream ofs{m_path + m_rd_name + "." + m_extension, std::ofstream::out};
+            // write header row
+            int c = 0;
+            ofs << "#";
+            ofs << "[" << c++ << "]step()";
+            ofs << m_sep;
+            ofs << "[" << c++ << "]time(s)";
+            ofs << m_sep;
+            ofs << "[" << c++ << "]Charge (C)";
+            ofs << std::endl;
+            // close file
+            ofs.close();
+        }
+    }
+}
+// end constructor
+
+// function that computes the charge at the surface of the EB
+void ChargeOnEB::ComputeDiags (int step)
+{
+    // Judge whether the diags should be done
+    if (!m_intervals.contains(step+1)) { return; }
+
+#if ((defined WARPX_DIM_3D) && (defined AMREX_USE_EB))
+    // get a reference to WarpX instance
+    auto & warpx = WarpX::GetInstance();
+
+    // Only compute the integral on level 0
+    int const lev = 0;
+
+    // get MultiFab data at lev
+    const MultiFab & Ex = warpx.getEfield(lev,0);
+    const MultiFab & Ey = warpx.getEfield(lev,1);
+    const MultiFab & Ez = warpx.getEfield(lev,2);
+
+    // get EB structures
+    amrex::EBFArrayBoxFactory const& eb_box_factory = warpx.fieldEBFactory(lev);
+    amrex::FabArray<amrex::EBCellFlagFab> const& eb_flag = eb_box_factory.getMultiEBCellFlagFab();
+    amrex::MultiCutFab const& eb_bnd_cent = eb_box_factory.getBndryCent();
+    amrex::MultiCutFab const& eb_bnd_normal = eb_box_factory.getBndryNormal();
+    amrex::Array<const amrex::MultiCutFab*,AMREX_SPACEDIM> eb_area_fraction = eb_box_factory.getAreaFrac();
+
+    // get surface integration element
+    const auto& warpx_instance = WarpX::GetInstance();
+    const auto dx = warpx_instance.Geom(lev).CellSizeArray();
+    amrex::Real const dSx = dx[1]*dx[2];
+    amrex::Real const dSy = dx[2]*dx[0];
+    amrex::Real const dSz = dx[0]*dx[1];
+
+    // Required for parser
+    const RealBox& real_box = warpx_instance.Geom(lev).ProbDomain();
+    const bool do_parser_weighting = m_do_parser_weighting;
+    auto fun_weightingparser =
+            utils::parser::compileParser<3>(m_parser_weighting.get());
+
+    // Integral to calculate
+    amrex::Gpu::Buffer<amrex::Real> surface_integral({0.0_rt});
+    amrex::Real* surface_integral_pointer = surface_integral.data();
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+    // Loop over boxes
+    for (amrex::MFIter mfi(Ex, TilingIfNotGPU()); mfi.isValid(); ++mfi)
+    {
+        const amrex::Box & box = enclosedCells(mfi.nodaltilebox());
```

Comment:
Could there be any problems related to staggering here? In other words, is it always safe to use a nodal tilebox?